_posts/2018-10-11-migrating-from-galen-to-backstop.md

```diff
@@ -0,0 +1,257 @@
+
+Therefore we decided to keep the files on the pipeline server, where the whole environment is constant.
+
+Sidenote: we still can run the tests locally. For that it is required to:
+
+1. Make the reference files from `master` branch.
+1. Apply whatever changes in the code.
+1. Run the test(s), and the images created will be checked against the ones created on the same OS.
+
+### 2. Large amount of files
+
+Right now, our reference files weight ~85MB. Let's say something gets changed in the header of the application. Most of the images would have to be refreshed and, say, ~60MB of new ones would be created. What would that mean? Downloading that ~60MB of data on next `git pull`, or whooping ~145MB of data on `git clone`. What if more changes are applied, or reverted? More and more MBs to download. That would be just too much.
+
+Sure, there are ways of bypassing this. One of the solutions is to use [LFS](https://git-lfs.github.com/). Yet, using that extension doesn't fix the no. 1 problem (rendering). We could go a bit further and use [Docker](https://www.docker.com/) to ensure we have the very same environment everywhere, but that just adds too much complexity 'only' for running visual regression tests.
+
+## Final words, learnings, possible improvements
+
+Occasionally we had, and still have, some tests failing from time to time (lazy loaded images, very rarely), but we are satisfied now. The most important thing is that we can ship more often. **We went down from ~30min to less than 5 with our pipeline!**
+
+As we have previously tested our application against various devices, we came to the conclusion, that perhaps we don't need to be so precise. Maybe it is OK to have it tested against the most popular browser out there (where most of the modern browsers behave almost identically) and on one device. If something will be moved 3 pixels to the right in some other browser / device - then so be it. The most important thing is that it is there, and that is what those tests are all about.
+
+The most important thing is that the time (30min down to 5) and the money saved (from the time to release and dropping of virtual machines from external partners) can now be used elsewhere, e.g. to provide more features to our customers.
+
+There are couple of things that can still be improved. One of them is not taking the screenshot of the entire page, when you want to test some state change of one of the elements on the page. You can take a screenshot of that one element only. We could also remove some tests (thus being faster). But that is not the bottleneck right now. It has moved somewhere else, and now we have to find that one and deal with it.
```

Comment:
In order to capture just one element you still have to render the entire page in the browser. So I don’t understand how this could make the tests faster.

Reply:
By comparing smaller images is faster than comparing large ones. I guess that was not clearly explained. Do you think that would be enough if I added that?

Reply:
I need to check. Perhaps that assumption (which it is) is wrong. I am just assuming that as less pixels is less work.

Reply:
Also I wouldn't call that a bottleneck (rendering part is probably much slower, or just the fact that that there are a lot of tests still). I wanted to point out that this could be an improvement.

Reply:
They are using `node-resemble-js`. See the file: https://github.com/garris/BackstopJS/blob/master/core/util/compare/compare-resemble.js